lib/mongo/cluster.rb

```diff
@@ -66,7 +66,7 @@ def ==(other)
     # @since 2.0.0
     def add(host)
       address = Address.new(host)
-      unless addresses.include?(address)
+      unless addresses.include?(address) || (@topology.single? && address.seed != @topology.seed)
```

Comment:
I would probable change the logic to to make it a bit less confusing as well as extracting a method on the second part...

```ruby
if !addresses.include?(address) || direct_connection?(address)
  #...
end
```